crates/transpiler/src/commutation_checker.rs

```
@@ -121,6 +126,137 @@ const fn build_supported_rotations() -> [Option<Option<StandardGate>>; STANDARD_
+/// Behavior is undefined if ``qubits`` contains a qubit index larger equal to ``num_qubits``,
+/// or if the number of Paulis in ``pauli_gate`` does not match the number of indices in ``qubits``.
+fn extract_op_from_pauligate(
+    pauli_gate: &OperationRef,
+    qubits: &[Qubit],
+    num_qubits: u32,
+) -> SparseObservable {
+    let OperationRef::Gate(gate) = pauli_gate else {
+        panic!("Invalid gate called 'pauli'");
+    };
+    let pauli_string = Python::attach(|py| {
+        let params = gate
+            .gate
+            .getattr(py, intern!(py, "params"))
+            .expect("Failed to read params.");
+
+        let params_list = params
+            .bind(py)
+            .cast::<PyList>()
+            .expect("Failed casting params to list");
+
+        params_list
+            .get_item(0)
+            .expect("Failed getting param content.")
+            .cast::<PyString>()
+            .expect("Failed casting to PyString")
+            .to_string()
+    });
+
+    let mut indices: Vec<u32> = Vec::new();
+    let mut bit_terms: Vec<BitTerm> = Vec::new();
+    for (i, pauli) in pauli_string.chars().rev().enumerate() {
+        match pauli {
+            'X' => {
+                indices.push(i as u32);
+                bit_terms.push(BitTerm::X)
+            }
+            'Y' => {
+                indices.push(i as u32);
+                bit_terms.push(BitTerm::Y)
+            }
+            'Z' => {
+                indices.push(i as u32);
+                bit_terms.push(BitTerm::Z)
+            }
+            _ => {} // nothing to do for the identity
+        }
+    }
+    // TODO there's the all-identity edge case we should potentially handle here,
+    // not sure is it's a valid state to otherwise end up with [0, 0] boundaries
+    let boundaries = vec![0, indices.len()];
+    let coeffs = vec![Complex64::new(1., 0.)];
+    let local =
+        SparseObservable::new(gate.num_qubits(), coeffs, bit_terms, indices, boundaries).unwrap();
```

why unwrap this here if the error could just be propagated upstream to the :meth:`SparseObservable::new(..)` method using '?', because I see that the method is defined as follows:
```rust 
pub fn new(
    num_qubits: u32,
    coeffs: Vec<Complex64>,
    bit_terms: Vec<BitTerm>,
    indices: Vec<u32>,
    boundaries: Vec<usize>,
) -> Result<Self, CoherenceError>
```
and hence should be able to handle errors, more gracefully than an unwrap. Is there a reason behind using unwrap here that I seem to be missing?

---

We could propagate the error upward all the way to the user-facing `commute` function and convert the error type, but I don't think that this `new` here can fail since we build the internal data manually and the indices are sorted. We could change this to a `new_unchecked` and avoid the checks and dealing with the result type here.

---

Changed to `new_unchecked` 👍🏻